app/models/qa/iri_template/variable_map.rb

```diff
@@ -0,0 +1,79 @@
+      def initialize(map)
+        @variable = extract_variable(config: map)
+        @required = extract_required(config: map)
+        @default = extract_default(config: map)
+        @property = map.fetch(:property, 'hydra:freetextQuery')
+      end
+
+      # Is this variable required?
+      # @returns true if required; otherwise, false
+      def required?
+        @required
+      end
+
+      # TODO: When implementing more complex query substitution, simple_value is used when template url specifies variable as {var_name}.
+      # Value to use in substitution, using default if one isn't passed in
+      # @param [Object] value to use if it exists
+      # @returns the value to use (e.g. 'fr')
+      def simple_value(sub_value = nil)
+        return sub_value.to_s if sub_value.present?
+        return default if default.present?
+        raise Qa::IriTemplate::MissingParameter, "#{variable} is required, but missing" if required?
+        ''
+      end
+
+      # TODO: When implementing more complex query substitution, parameter_value is used when template url specifies variable as {?var_name}.
+      # # Parameter and value to use in substitution, using default is one isn't passed in
+      # # @param [Object] value to use if it exists
+      # # @returns the parameter and value to use (e.g. 'language=fr')
+      # def parameter_value(sub_value = nil)
+      #   simple_value = simple_value(sub_value)
+      #   return '' if simple_value.blank?
+      #   param_value = "#{variable}=#{simple_value}"
+      # end
+
+      private
+
+        # Extract the variable name from the config
+        # @param config [Hash] configuration (json) holding the variable map
+        # @param var [Symbol] key identifying the variable in the configuration
+        # @return [String] variable for substitution in the url tmeplate
+        def extract_variable(config:, var: :variable)
+          varname = config.fetch(var, nil)
+          raise Qa::InvalidConfiguration, 'variable is required' unless varname
+          varname
+        end
+
+        # Extract the variable name from the config
+        # @param config [Hash] configuration (json) holding the variable map
+        # @param var [Symbol] key in the configuration identifying whether the variable is required
+        # @return [True | False] required as true or false
+        def extract_required(config:, var: :required)
+          required = config.fetch(var, nil)
+          raise Qa::InvalidConfiguration, 'required must be true or false' unless required == true || required == false
+          required
+        end
+
+        # Extract the default value from the config ignoring defaults for required variables
+        # @param config [Hash] configuration (json) holding the variable map
+        # @param var [Symbol] key identifying the default value in the configuration
+        # @return [String] default value to use for the variable; nil if variable is required
```

Comment:
This comment does not match the code.  Nil is not returned only empty string.  You are taking care of the nil in simple value, although I like the idea of taking care of it down here.  It would simplify simple value.

Reply:
Good catch.  I adjusted the processing of the default value in general.  The exception in simple_value method for missing required is thrown before the extract_default method is called since it doesn't apply for required variables.